jp/lessons/specifics/ets.md

```diff
@@ -0,0 +1,299 @@
+---
+layout: page
+title: Erlang Term Storage (ETS)
+category: specifics
+order: 4
+lang: jp
+---
+
+Erlang Term Storage、略してETSは、OTP内部に組み込まれている強力なストレージエンジンで、Elixirから利用することができます。このレッスンではETSを呼び出す方法と、アプリケーションで使用する方法を見ていきます。
+
+## 目次
+
+- [概要](#section-1)
+- [テーブルの作成](#section-2)
+  - [テーブルの型](#section-3)
+  - [アクセス制御](#section-4)
+- [データの挿入](#section-5)
+- [データの復旧](#section-6)
+  - [キー探索](#section-7)
+  - [単純なマッチ](#section-8)
+  - [発展的な探索](#section-9)
+- [データの削除](#seciton-10)
+  - [レコードの除去](#section-11)
+  - [テーブルの除去](#section-12)
+- [ETSの使用例](#ets)
+- [ディスクを用いたETS](#ets-1)
+
+## 概要
+
+ETSは堅牢なインメモリストアで、ElixirやErlangのオブジェクトを保管するためのものです。ETSはとても大きなデータを格納し、定数時間でのデータアクセスを提供します。
+
+ETSのテーブルは個々のプロセスによって生成、所有されています。テーブルを所有しているプロセスが死ぬと、テーブルは破壊されます。初期状態では、ETSはノードにつき1400テーブルに制限されています。
+
+## テーブルの作成
+
+テーブルは`new/2`で作成します。この関数はテーブル名と、一連のオプションを受け取り、以降の操作で用いることのできるテーブルのIDを返します。
+
+例としてユーザをニックネームで管理し、検索するテーブルを作ります:
+
+```elixir
+iex> table = :ets.new(:user_lookup, [:set, :protected])
+8212
+```
+
+GenServerと同じように、IDの代わりに名前でETSテーブルにアクセスする方法があります。`:named_table`をオプションに含めることで、名前を用いて直接テーブルにアクセスできます:
+
+```elixir
+iex> :ets.new(:user_lookup, [:set, :protected, :named_table])
+:user_lookup
+```
+
+### テーブルの型
+
+ETSで利用できるテーブルの型は4つあります:
+
++ `set` - デフォルトのテーブル型です。キー毎に1つの値を持ち、キーは一意です。
++ `ordered_set` - `set`に似ていますがErlang/Elixirの条件で順序付けされます。重要なので注記しておくと、キーの比較は`ordered_set`の内部では他と異なります。等値であるとみなされるキーは個別にマッチしません。1と1.0は同じキーであるとみなされます。
++ `bag` - キー毎に複数のオブジェクトを持てるが、1つのオブジェクトにつき1つのインスタンスのみです。
++ `duplicate_bag` - キー毎に複数のオブジェクトが持て、重複が許されます。
+
+### アクセス制御
+
+ETSでのアクセス制御はモジュール内部のアクセス制御と似ています:
+
++ `public` - 全てのプロセスで読み/書きが可能です。
++ `protected` - 全てのプロセスで読み込みが可能です。所有プロセスによってのみ書き込みができます。これがデフォルトです。
++ `private` - 読み/書きは所有プロセスに限定されます。
+
+## データの挿入
+
+ETSはスキーマを持ちません。唯一の制限は、データは最初の項がキーであるタプルとして格納されなくてはいけないというものです。新しいデータを追加するには`insert/2`を用います:
+
+```elixir
+iex> :ets.insert(:user_lookup, {"doomspork", "Sean", ["Elixir", "Ruby", "Java"]})
+true
+```
+
+`insert/2`を`set`や`ordered_set`に用いると、既存のデータは置換されます。これを避けるため、キーが存在している場合に`false`を返してくれる`insert_new/2`があります:
+
+```elixir
+iex> :ets.insert_new(:user_lookup, {"doomspork", "Sean", ["Elixir", "Ruby", "Java"]})
+false
+iex> :ets.insert_new(:user_lookup, {"3100", "", ["Elixir", "Ruby", "JavaScript"]})
```

Comment:
You can fill in your name for the second value if you want, I couldn't glean it from any of your online profiles :grinning: